client/txn.go

```diff
@@ -406,12 +406,21 @@ func (txn *Txn) send(calls ...proto.Call) error {
 		}
 	}
 
-	_, haveEndTxn := lastReq.(*proto.EndTransactionRequest)
+	endTxnRequest, haveEndTxn := lastReq.(*proto.EndTransactionRequest)
 	needEndTxn := txn.Proto.Writing || haveTxnWrite
+	elideEndTxn := haveEndTxn && !needEndTxn
 
-	if haveEndTxn && !needEndTxn {
+	if elideEndTxn {
 		calls = calls[:lastIndex]
 	}
 
-	return txn.db.send(calls...)
+	err := txn.db.send(calls...)
+	if elideEndTxn && err == nil {
+		if endTxnRequest.Commit {
```

Comment:
added a comment and some tests; sadly had to touch the test infrastructure more than I would have liked.